File: rate_limiter.c
/* Deterministic contract lowered from the reviewed V2 domain 'rate_limiter'.
 * Human review of the reviewed artifact is required before trust. */

typedef struct {
    int tokens;
} rate_limiter;

/*@
  requires \valid(counter);
  assigns counter->tokens;
  ensures counter->tokens == 5;
  ensures (0 <= counter->tokens) && (counter->tokens <= 5);
*/
void rate_limiter_init(rate_limiter *counter) {
    counter->tokens = 5;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->tokens;
*/
int rate_limiter_get_tokens(const rate_limiter *counter) {
    return counter->tokens;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->tokens) && (counter->tokens <= 5);
  requires counter->tokens > 0;
  assigns counter->tokens;
  ensures counter->tokens == \old(counter->tokens) - 1;
  ensures (0 <= counter->tokens) && (counter->tokens <= 5);
*/
void rate_limiter_handle_request(rate_limiter *counter) {
    int pre_tokens = counter->tokens;
    counter->tokens = pre_tokens - 1;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->tokens) && (counter->tokens <= 5);
  requires counter->tokens < 5;
  assigns counter->tokens;
  ensures counter->tokens == 5;
  ensures (0 <= counter->tokens) && (counter->tokens <= 5);
*/
void rate_limiter_refill(rate_limiter *counter) {
    counter->tokens = 5;
}

File: RateLimiter.rs
use prusti_contracts::*;

pub struct RateLimiter {
    pub tokens: i32,
}

impl RateLimiter {
    #[ensures(result.tokens == 5 && (0 <= result.tokens) && (result.tokens <= 5))]
    pub fn new() -> Self {
        Self { tokens: 5 }
    }

    #[pure]
    #[ensures(result == self.tokens)]
    pub fn get_tokens(&self) -> i32 {
        self.tokens
    }

    #[requires((0 <= self.tokens) && (self.tokens <= 5))]
    #[requires(self.tokens > 0)]
    #[ensures(self.tokens == old(self.tokens) - 1)]
    #[ensures((0 <= self.tokens) && (self.tokens <= 5))]
    pub fn handle_request(&mut self) {
        let pre_tokens = self.tokens;
        self.tokens = pre_tokens - 1;
    }

    #[requires((0 <= self.tokens) && (self.tokens <= 5))]
    #[requires(self.tokens < 5)]
    #[ensures(self.tokens == 5)]
    #[ensures((0 <= self.tokens) && (self.tokens <= 5))]
    pub fn refill(&mut self) {
        self.tokens = 5;
    }
}
